Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,177 @@
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> GetAsync(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                this.log.Error("Device model id cannot be empty!", () => {});
+                throw new InvalidInputException("Device model id cannot be empty! ");
+            }
+
+            var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel, bool generateId = true)
+        {
+            deviceModel.Created = DateTimeOffset.UtcNow;
+            deviceModel.Modified = deviceModel.Created;
+
+            if (generateId)
+            {
+                deviceModel.Id = Guid.NewGuid().ToString();
+            }
+
+            this.log.Debug("Creating a custom device model.", () => new { deviceModel });
+
+            // Note: using UpdateAsync because the service generates the ID
+            var result = await this.storage.UpdateAsync(
+                STORAGE_COLLECTION,
+                deviceModel.Id,
+                JsonConvert.SerializeObject(deviceModel),
+                null);
+
+            deviceModel.ETag = result.ETag;
+
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create or replace a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> UpsertAsync(DeviceModel deviceModel)
+        {
+            var id = deviceModel.Id;
+            var eTag = deviceModel.ETag;
+
+            try
+            {
+                var item = await this.GetAsync(id);
+
+                if (item.ETag == eTag)
+                {
+                    // Replace a custom device model
+                    deviceModel.Created = item.Created;
+                    deviceModel.Modified = DateTimeOffset.UtcNow;
+
+                    this.log.Debug("Modifying a custom device model via PUT.", () => new { deviceModel });
+
+                    var result = await this.storage.UpdateAsync(
+                        STORAGE_COLLECTION,
+                        id,
+                        JsonConvert.SerializeObject(deviceModel),
+                        eTag);
+
+                    // Return the new ETag provided by the storage
+                    deviceModel.ETag = result.ETag;
+                }
+                else
+                {
+                    this.log.Error("Invalid ETag.", () => new { CurrentETag = item.ETag, ETagProvided = eTag });
+                    throw new ConflictingResourceException("Invalid ETag. Device Model ETag is:'" + item.ETag + "'.");
```

Comment:
see my previous comment, this will be caught at line 163. Shouldn't it pop up and result in some HTTP response error?

Reply:
suggestion: before changing the code, try to repro my scenario with a unit test. That will give you 100% confidence

Reply:
Done.